consensus/tendermint/core/backlog.go

```diff
@@ -75,59 +75,65 @@ func (c *core) storeBacklog(msg *Message, src types.CommitteeMember) {
-		var isFuture bool
-
-		// We stop processing if
-		//   1. backlog is empty
-		//   2. The first message in queue is a future message
-		for !(backlog.Empty() || isFuture) {
-			m, prio := backlog.Pop()
-			msg := m.(*Message)
-			msgRound, _ := msg.Round() // error checking done before push
-			msgHeight, _ := msg.Height()
-
-			// Push back if it's a future message
-			err := c.checkMessage(msgRound, msgHeight, Step(msg.Code))
-			if err != nil {
-				if err == errFutureHeightMessage || err == errFutureRoundMessage || err == errFutureStepMessage {
-					logger.Debug("Stop processing backlog", "msg", msg, "err", err)
-					backlog.Push(msg, prio)
-					isFuture = true
-					break
+
+		initialLen := len(backlog)
+		if initialLen > 0 {
+			// For loop will change the size for backlog therefore we need to keep track of the initial length and
+			// adjust for index change. This is done by keeping track of how many elements have been removed and
+			// subtracting it from the for-loop iterator, since each removed element will cause the index to change for
+			// each element after the removed element.
+			//
+			// If the message is a future height, round or step message then the for-loop will move on to the next
+			// iteration, however, if any other error occurs then the message is removed from the backlog and for-loop
+			// moves to the next iteration.
+			//
+			// If there are no errors when checking the message then the message is sent to the handler via a goroutine
+			// and this message is removed from the backlog.
+			totalElemRemoved := 0
+			for i := 0; i < initialLen; i++ {
+				offset := i - totalElemRemoved
+				curMsg := backlog[offset]
+
+				r, _ := curMsg.Round()
+				h, _ := curMsg.Height()
+				if err := c.checkMessage(r, h, Step(curMsg.Code)); err != nil {
+					if err == errFutureHeightMessage || err == errFutureRoundMessage || err == errFutureStepMessage {
+						logger.Debug("Futrue message in backlog", "msg", curMsg, "err", err)
+						continue
+					}
+					logger.Debug("Skipping the backlog message", "msg", curMsg, "err", err)
+				} else {
+					logger.Debug("Post backlog event", "msg", curMsg)
+
+					go c.sendEvent(backlogEvent{
+						src: *c.lastHeader.CommitteeMember(src),
+						msg: curMsg,
+					})
 				}
-				logger.Debug("Skip the backlog event", "msg", msg, "err", err)
-				continue
-			}
-			logger.Debug("Post backlog event", "msg", msg)
 
-			go c.sendEvent(backlogEvent{
-				src: src,
-				msg: msg,
-			})
+				backlog = append(backlog[:offset], backlog[offset+1:]...)
+				totalElemRemoved++
+			}
+			// We need to ensure that there is no memory leak by reallocating new memory if the original underlying
+			// array become very large and only a small part of it is being used by the slice. We need len(backlog) > 0
+			// check again since the backlog size can change and cause division by zero errors.
+			if len(backlog) > 0 && cap(backlog)/len(backlog) > capToLenRatio {
```

Comment:
```suggestion
			if len(backlog) == 0 || cap(backlog)/len(backlog) > capToLenRatio {
```

This covers the case when the backlog is empty

Reply:
Surely `if len(backlog) == 0` condition can cause division by zero errors, or am I missing something? 

Its possible to have `len(s) == 0` and `cap(s) == 0` at the same time, while it is impossible to have `len(s)>0` and `cap(s)==0` (at the same time). I only want to check this condition when backlog is not empty.

Reply:
To free the memory held by the backlog. As it stands when `len(backlog) == 0` we don't free anything, regardless of the magnitude of `cap(backlog)`.

In fact I think this would be a better approach.

```suggestion
			if cap(backlog)/capToLenRatio > len(backlog) {
```
It solves the division by 0 issue.

Reply:
Very nice catch! Forgot to think about the `len(s)==0 and cap(backlog) >> len(s)`. could you please also update the comment in your suggestion